File: pyFull/scripts/sources/odds_batch_historical.py
import os
import json
import time
import math
import re
import requests
from urllib.parse import quote
from datetime import datetime, timedelta

_dir = os.path.dirname(os.path.abspath(__file__))
ODDS_CACHE_DIR = os.path.join(_dir, "..", "..", "..", "data", "odds_cache", "nba")

BASE = "https://api.the-odds-api.com/v4"


def _sleep(ms):
    time.sleep(ms / 1000.0)


def norm_key(s):
    return re.sub(r"\s+", " ", re.sub(r"[^a-z0-9\s]", " ", str(s or "").lower())).strip()


def norm_team(name):
    return re.sub(r"\s+", " ", str(name or "").strip())


TEAM_ALIASES = {
    "la lakers": ["los angeles lakers", "la lakers"],
    "la clippers": ["los angeles clippers", "la clippers"],
    "golden state": ["golden state warriors", "golden state"],
    "oklahoma city": ["oklahoma city thunder", "oklahoma city"],
    "new orleans": ["new orleans pelicans", "new orleans"],
    "new york": ["new york knicks", "new york"],
    "san antonio": ["san antonio spurs", "san antonio"],
    "portland": ["portland trail blazers", "portland"],
    "philadelphia": ["philadelphia 76ers", "philadelphia"],
}


def expand_aliases(name):
    k = norm_key(name)
    return TEAM_ALIASES.get(k, [k])


def pick_bookmaker(bookmakers):
    if not isinstance(bookmakers, list) or len(bookmakers) == 0:
        return None
    preferred = ["DraftKings", "FanDuel", "BetMGM", "Caesars", "PointsBet", "BetRivers"]
    for title in preferred:
        found = next((b for b in bookmakers if (b or {}).get("title") == title), None)
        if found:
            return found
    return bookmakers[0]


def find_market(bookmaker, key):
    if not bookmaker or not bookmaker.get("markets"):
        return None
    return next((m for m in bookmaker["markets"] if (m or {}).get("key") == key), None)


def to_model_line(home_team, away_team, spread_points, team_for_spread):
    if not math.isfinite(spread_points):
        return None
    abs_val = abs(spread_points)
    is_home = team_for_spread == home_team
    is_away = team_for_spread == away_team
    if not is_home and not is_away:
        return None
    if is_home:
        return abs_val if spread_points < 0 else -abs_val
    return -abs_val if spread_points < 0 else abs_val


def fetch_with_retry(url, tries=5):
    wait = 700
    for i in range(tries):
        res = requests.get(url)
        if res.status_code != 429:
            return res
        print(f"  [odds_batch] 429 rate limited, retrying in {wait}ms...")
        _sleep(wait)
        wait = min(wait * 2, 8000)
    return requests.get(url)


# -- Fetch one snapshot from the API -------------------------------------------

def fetch_snapshot(api_key, ts):
    url = (
        f"{BASE}/historical/sports/basketball_nba/odds?"
        f"apiKey={quote(api_key)}"
        f"&regions=us"
        f"&markets=spreads,totals"
        f"&oddsFormat=american"
        f"&date={quote(ts)}"
    )

    res = fetch_with_retry(url)
    if res.status_code != 200:
        txt = res.text
        raise Exception(f"Historical odds fetch failed: {res.status_code} {txt}")

    json_data = res.json()
    data = json_data.get("data")
    return data if isinstance(data, list) else []


# -- Extract odds for one game from a snapshot ---------------------------------

def extract_odds_for_game(data, home, away):
    home_aliases = expand_aliases(home)
    away_aliases = expand_aliases(away)

    ev = next(
        (e for e in data
         if norm_key((e or {}).get("home_team")) in home_aliases
         and norm_key((e or {}).get("away_team")) in away_aliases),
        None
    )
    if not ev:
        ev = next(
            (e for e in data
             if any(ha in norm_key((e or {}).get("home_team", "")) or norm_key((e or {}).get("home_team", "")) in ha for ha in home_aliases)
             and any(aa in norm_key((e or {}).get("away_team", "")) or norm_key((e or {}).get("away_team", "")) in aa for aa in away_aliases)),
            None
        )

    if not ev:
        return None

    home_team = norm_team(ev.get("home_team"))
    away_team = norm_team(ev.get("away_team"))
    book = pick_bookmaker(ev.get("bookmakers"))

    line = None
    total = None

    spreads = find_market(book, "spreads") if book else None
    totals = find_market(book, "totals") if book else None

    if spreads and spreads.get("outcomes"):
        out = next((o for o in spreads["outcomes"] if math.isfinite(float((o or {}).get("point", "nan")))), None)
        if out:
            line = to_model_line(home_team, away_team, float(out["point"]), norm_team(out.get("name")))

    if totals and totals.get("outcomes"):
        out = next((o for o in totals["outcomes"] if math.isfinite(float((o or {}).get("point", "nan")))), None)
        if out:
            total = float(out["point"])

    return {"line": line, "total": total, "_book": (book or {}).get("title")}


# -- Main export: fetch ALL odds for a date in 1-2 API calls -------------------

def fetch_odds_for_day(date_yyyymmdd, games_list):
    """
    Fetch ALL odds for a date in 1-2 API calls instead of per-game.
    Uses disk cache so re-runs don't hit the API at all.
    Returns: dict of "away@home" -> { line, total, _book, _note }
    """
    api_key = os.environ.get("ODDS_API_KEY")

    # Check disk cache first
    if not os.path.exists(ODDS_CACHE_DIR):
        os.makedirs(ODDS_CACHE_DIR, exist_ok=True)
    cache_path = os.path.join(ODDS_CACHE_DIR, date_yyyymmdd + ".json")
    if os.path.exists(cache_path):
        with open(cache_path, "r", encoding="utf-8") as f:
            cached = json.load(f)
        # Fuzzy-match cache keys to ESPN names via alias expansion
        if games_list:
            cache_entries = list(cached.items())
            for g in games_list:
                espn_key = f"{g['away']}@{g['home']}"
                if espn_key in cached:
                    continue
                away_aliases = expand_aliases(g["away"])
                home_aliases = expand_aliases(g["home"])
                for ck, cv in cache_entries:
                    parts = ck.split("@")
                    if len(parts) == 2:
                        c_away, c_home = norm_key(parts[0]), norm_key(parts[1])
                        if c_away in away_aliases and c_home in home_aliases:
                            cached[espn_key] = cv
                            break
        print(f"  [odds_batch] {date_yyyymmdd}: loaded from cache ({len(cached)} games)")
        return cached

    if not api_key:
        raise Exception("Missing ODDS_API_KEY env var (no cache found, need API).")

    # Build snapshot timestamps to try
    date_str = f"{date_yyyymmdd[0:4]}-{date_yyyymmdd[4:6]}-{date_yyyymmdd[6:8]}"

    # Game-specific times: 90 min before each game's tipoff
    game_specific_times = []
    for g in games_list:
        commence = g.get("commenceTimeIso")
        if commence:
            try:
                d = datetime.fromisoformat(commence.replace("Z", "+00:00"))
                d = d - timedelta(minutes=90)
                game_specific_times.append(d.strftime("%Y-%m-%dT%H:%M:%SZ"))
            except (ValueError, AttributeError):
                pass

    # Fixed fallbacks: 4 PM UTC (11 AM ET) + 11 PM UTC (6 PM ET)
    fallbacks = [
        f"{date_str}T16:00:00Z",
        f"{date_str}T23:00:00Z",
    ]

    # Deduplicate: game-specific first, then fallbacks
    seen = set()
    all_times = []
    for t in game_specific_times + fallbacks:
        if t not in seen:
            seen.add(t)
            all_times.append(t)

    results = {}
    snapshot_data = None

    for ts in all_times:
        try:
            snapshot_data = fetch_snapshot(api_key, ts)
            _sleep(500)
        except Exception as e:
            print(f"  [odds_batch] Snapshot {ts} failed: {e}")
            continue

        if not snapshot_data or len(snapshot_data) == 0:
            continue

        # Extract odds for all games we're looking for
        found = 0
        for g in games_list:
            key = f"{g['away']}@{g['home']}"
            if key in results and isinstance(results[key].get("line"), (int, float)):
                continue  # already found

            odds = extract_odds_for_game(snapshot_data, g["home"], g["away"])
            if odds and isinstance(odds.get("line"), (int, float)):
                results[key] = {**odds, "_note": f"batch snapshot {ts}"}
                found += 1

        print(f"  [odds_batch] Snapshot {ts[11:16]} UTC: found {found} new games ({len(results)}/{len(games_list)} total)")

        # If we found all games, stop
        if len(results) >= len(games_list):
            break

    # Fill missing games with null
    for g in games_list:
        key = f"{g['away']}@{g['home']}"
        if key not in results:
            results[key] = {"line": None, "total": None, "_book": None, "_note": "Not found in any snapshot"}

    # Cache to disk
    try:
        with open(cache_path, "w", encoding="utf-8") as f:
            json.dump(results, f, indent=2)
    except Exception:
        pass  # non-critical

    return results

File: jsNBA/scripts/sources/odds_batch_historical.mjs
// sources/odds_batch_historical.mjs
// ────────────────────────────────────────────────────────────────────────────
// BATCH historical odds fetch — gets ALL games for a date in 1-2 API calls
// instead of per-game (10 games × 3 retries = 30 calls → 1-2 calls).
//
// Uses disk cache so re-runs don't hit the API at all.
// ────────────────────────────────────────────────────────────────────────────

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ODDS_CACHE_DIR = path.join(__dirname, "..", "..", "..", "data", "odds_cache", "nba");

const BASE = "https://api.the-odds-api.com/v4";

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

function normKey(s) {
  return String(s || "").toLowerCase().replace(/[^a-z0-9\s]/g, " ").replace(/\s+/g, " ").trim();
}

function normTeam(name) {
  return String(name || "").trim().replace(/\s+/g, " ");
}

const TEAM_ALIASES = {
  "la lakers":      ["los angeles lakers", "la lakers"],
  "la clippers":    ["los angeles clippers", "la clippers"],
  "golden state":   ["golden state warriors", "golden state"],
  "oklahoma city":  ["oklahoma city thunder", "oklahoma city"],
  "new orleans":    ["new orleans pelicans", "new orleans"],
  "new york":       ["new york knicks", "new york"],
  "san antonio":    ["san antonio spurs", "san antonio"],
  "portland":       ["portland trail blazers", "portland"],
  "philadelphia":   ["philadelphia 76ers", "philadelphia"],
};

function expandAliases(name) {
  const k = normKey(name);
  return TEAM_ALIASES[k] || [k];
}

function pickBookmaker(bookmakers) {
  if (!Array.isArray(bookmakers) || bookmakers.length === 0) return null;
  const preferred = ["DraftKings", "FanDuel", "BetMGM", "Caesars", "PointsBet", "BetRivers"];
  for (const title of preferred) {
    const found = bookmakers.find(b => b?.title === title);
    if (found) return found;
  }
  return bookmakers[0];
}

function findMarket(bookmaker, key) {
  if (!bookmaker?.markets) return null;
  return bookmaker.markets.find(m => m?.key === key) || null;
}

// Sportsbook convention: -X = home favored by X, +X = away favored by X
function toModelLine(homeTeam, awayTeam, spreadPoints, teamForSpread) {
  if (!Number.isFinite(spreadPoints)) return null;
  const isHome = teamForSpread === homeTeam;
  const isAway = teamForSpread === awayTeam;
  if (!isHome && !isAway) return null;
  if (isHome) return spreadPoints;
  return -spreadPoints;
}

async function fetchWithRetry(url, tries = 5) {
  let wait = 700;
  for (let i = 0; i < tries; i++) {
    const res = await fetch(url);
    if (res.status !== 429) return res;
    console.warn(`  [odds_batch] 429 rate limited, retrying in ${wait}ms...`);
    await sleep(wait);
    wait = Math.min(wait * 2, 8000);
  }
  return fetch(url);
}

// ── Fetch one snapshot from the API ─────────────────────────────────────────

async function fetchSnapshot(apiKey, ts) {
  const url =
    `${BASE}/historical/sports/basketball_nba/odds?` +
    `apiKey=${encodeURIComponent(apiKey)}` +
    `&regions=us` +
    `&markets=spreads,totals` +
    `&oddsFormat=american` +
    `&date=${encodeURIComponent(ts)}`;

  const res = await fetchWithRetry(url);
  if (!res.ok) {
    const txt = await res.text().catch(() => "");
    throw new Error(`Historical odds fetch failed: ${res.status} ${txt}`);
  }

  const json = await res.json();
  return Array.isArray(json?.data) ? json.data : [];
}

// ── Extract odds for one game from a snapshot ───────────────────────────────

function extractOddsForGame(data, home, away) {
  const homeAliases = expandAliases(home);
  const awayAliases = expandAliases(away);

  const ev =
    data.find(e => {
      const h = normKey(e?.home_team);
      const a = normKey(e?.away_team);
      return homeAliases.includes(h) && awayAliases.includes(a);
    }) ||
    data.find(e => {
      const h = normKey(e?.home_team);
      const a = normKey(e?.away_team);
      return homeAliases.some(ha => h.includes(ha) || ha.includes(h)) &&
             awayAliases.some(aa => a.includes(aa) || aa.includes(a));
    });

  if (!ev) return null;

  const homeTeam = normTeam(ev.home_team);
  const awayTeam = normTeam(ev.away_team);
  const book = pickBookmaker(ev?.bookmakers);

  let line = null, total = null;

  const spreads = book ? findMarket(book, "spreads") : null;
  const totals = book ? findMarket(book, "totals") : null;

  if (spreads?.outcomes?.length) {
    const out = spreads.outcomes.find(o => Number.isFinite(Number(o?.point)));
    if (out) line = toModelLine(homeTeam, awayTeam, Number(out.point), normTeam(out.name));
  }

  if (totals?.outcomes?.length) {
    const out = totals.outcomes.find(o => Number.isFinite(Number(o?.point)));
    if (out) total = Number(out.point);
  }

  return { line, total, _book: book?.title ?? null };
}

// ── Main export: fetch ALL odds for a date in 1-2 API calls ─────────────────
//
// Strategy: fetch a snapshot at a time when all games should have lines posted
// but none have tipped off yet. For NBA that's ~5:00 PM ET (22:00 UTC) on
// game day. If any games are missing, try an earlier snapshot.
//
// Returns: Map of "away@home" → { line, total, _book, _note }

export async function fetchOddsForDay(dateYYYYMMDD, gamesList) {
  const apiKey = process.env.ODDS_API_KEY;

  // Check disk cache first
  if (!fs.existsSync(ODDS_CACHE_DIR)) fs.mkdirSync(ODDS_CACHE_DIR, { recursive: true });
  const cachePath = path.join(ODDS_CACHE_DIR, dateYYYYMMDD + ".json");
  if (fs.existsSync(cachePath)) {
    const cached = JSON.parse(fs.readFileSync(cachePath, "utf8"));
    // Fuzzy-match cache keys to ESPN names via alias expansion
    if (gamesList?.length) {
      const cacheEntries = Object.entries(cached);
      for (const g of gamesList) {
        const espnKey = `${g.away}@${g.home}`;
        if (cached[espnKey]) continue;
        const awayAliases = expandAliases(g.away);
        const homeAliases = expandAliases(g.home);
        for (const [ck, cv] of cacheEntries) {
          const [cAway, cHome] = ck.split("@").map(normKey);
          if (awayAliases.includes(cAway) && homeAliases.includes(cHome)) {
            cached[espnKey] = cv;
            break;
          }
        }
      }
    }
    console.log(`  [odds_batch] ${dateYYYYMMDD}: loaded from cache (${Object.keys(cached).length} games)`);
    return cached;
  }

  if (!apiKey) throw new Error("Missing ODDS_API_KEY env var (no cache found, need API).");

  // Build snapshot timestamps to try.
  // Priority: per-game commence-based times FIRST (90 min before tipoff),
  // then fixed fallbacks. This avoids grabbing live/in-game lines for
  // early-start games (e.g. Sunday noon tips).
  const dateStr = `${dateYYYYMMDD.slice(0,4)}-${dateYYYYMMDD.slice(4,6)}-${dateYYYYMMDD.slice(6,8)}`;

  // Game-specific times: 90 min before each game's tipoff
  const gameSpecificTimes = [];
  for (const g of gamesList) {
    if (g.commenceTimeIso) {
      const d = new Date(g.commenceTimeIso);
      if (!Number.isNaN(d.getTime())) {
        d.setMinutes(d.getMinutes() - 90);
        gameSpecificTimes.push(d.toISOString().replace(/\.\d{3}Z$/, "Z"));
      }
    }
  }

  // Fixed fallbacks (only used if commence times aren't available)
  const fallbacks = [
    `${dateStr}T16:00:00Z`,  // 4 PM UTC — 11 AM ET, catches early tips
    `${dateStr}T23:00:00Z`,  // 11 PM UTC — 6 PM ET, catches evening tips
  ];

  // Deduplicate: game-specific first, then fallbacks
  const allTimes = [...new Set([...gameSpecificTimes, ...fallbacks])];

  const results = {};
  let snapshotData = null;

  for (const ts of allTimes) {
    try {
      snapshotData = await fetchSnapshot(apiKey, ts);
      await sleep(500);  // one sleep between snapshot calls
    } catch (e) {
      console.warn(`  [odds_batch] Snapshot ${ts} failed: ${e.message}`);
      continue;
    }

    if (!snapshotData || snapshotData.length === 0) continue;

    // Extract odds for all games we're looking for
    let found = 0;
    for (const g of gamesList) {
      const key = `${g.away}@${g.home}`;
      if (results[key] && typeof results[key].line === "number") continue; // already found

      const odds = extractOddsForGame(snapshotData, g.home, g.away);
      if (odds && typeof odds.line === "number") {
        results[key] = { ...odds, _note: `batch snapshot ${ts}` };
        found++;
      }
    }

    console.log(`  [odds_batch] Snapshot ${ts.slice(11, 16)} UTC: found ${found} new games (${Object.keys(results).length}/${gamesList.length} total)`);

    // If we found all games, stop
    if (Object.keys(results).length >= gamesList.length) break;
  }

  // Fill missing games with null
  for (const g of gamesList) {
    const key = `${g.away}@${g.home}`;
    if (!results[key]) {
      results[key] = { line: null, total: null, _book: null, _note: "Not found in any snapshot" };
    }
  }

  // Cache to disk
  try { fs.writeFileSync(cachePath, JSON.stringify(results, null, 2)); }
  catch (e) { /* non-critical */ }

  return results;
}
